core/src/Streamly/Internal/FileSystem/Posix/ReadDir.hsc

```diff
@@ -138,41 +257,64 @@ isMetaDir dname = do
                 then return True
                 else return False
 
-{-# NOINLINE lstatDname #-}
-lstatDname :: PosixPath -> Ptr CChar -> IO (Bool, Bool)
-lstatDname parent dname = do
+data GStatRes
+    = GSIsMetaDir
+    | GSIsRegDir
+    | GSIsNotDir
+    | GSIgnoreError Errno
+
+{-# NOINLINE gstatDname #-}
+gstatDname
+    :: ReadOptions -> PosixPath -> Ptr CChar -> IO GStatRes
+gstatDname conf parent dname = do
     isMeta <- liftIO $ isMetaDir dname
     if isMeta
-    then pure (True, True)
+    then pure GSIsMetaDir
     else do
         -- XXX We can create a pinned array right here since the next call pins
         -- it anyway.
         path <- appendCString parent dname
         Array.asCStringUnsafe (Path.toChunk path) $ \cStr -> do
-            res <- c_lstat_is_directory cStr
+            res <-
+                if _followSymlinks conf
+                then statWith c_stat cStr
+                else statWith c_lstat cStr
             case res of
-                x | x == 1 -> pure (True, False)
-                x | x == 0 -> pure (False, False)
-                -- XXX Need to check if and how we should handle some errors
-                -- like EACCES.
-                _ -> throwErrno "checkIfDirectory"
+                Left errno ->
+                    if    (errno == eNOENT && _ignoreNonExistingFiles conf)
+                       || (errno == eACCES && _ignoreInAccessibleFiles conf)
+                       || (errno == eLOOP && _ignoreSymlinkLoopErrors conf)
+                    then pure $ GSIgnoreError errno
+                    else throwErrno (errMsg path)
+                Right mode -> pure $
+                    if (mode == s_IFDIR)
+                    then GSIsRegDir
+                    else GSIsNotDir
+    where
+    errMsg path =
+        let pathStr = Path.toString path
+         in "statDname: " ++ pathStr
 
 -- | Checks if dname is a directory and additionaly returns if dname is a meta
 -- directory.
 {-# INLINE checkDirStatus #-}
 checkDirStatus
-    :: PosixPath -> Ptr CChar -> #{type unsigned char} -> IO (Bool, Bool)
+    :: ReadOptions
+    -> PosixPath -> Ptr CChar -> #{type unsigned char} -> IO GStatRes
 #ifdef FORCE_LSTAT_READDIR
-checkDirStatus parent dname _ = lstatDname parent dname
+checkDirStatus conf parent dname _ =
+    gstatDname conf parent dname
 #else
-checkDirStatus parent dname dtype =
+checkDirStatus conf parent dname dtype =
     if dtype == (#const DT_DIR)
     then do
         isMeta <- liftIO $ isMetaDir dname
-        pure (True, isMeta)
-    else if dtype /= #const DT_UNKNOWN
-         then pure (False, False)
-         else lstatDname parent dname
+        pure $ if isMeta then GSIsMetaDir else GSIsRegDir
+    else if dtype == (#const DT_LNK) && _followSymlinks conf
```

Comment:
keep _followSymlinks inside the if statement instead of using `&&`.

Reply:
Any particular reason why?

Reply:
Same reason as the errno case earlier. If dtype == (#const DT_LNK) is True and _followSymlinks conf is False we still go check the else case but we should not, we know it cannot be DT_UNKNOWN.

Reply:
Now that I look at it again. What was written is incorrect logic. Will push a fix.